pkg/experimental/make_gen_rpm.py

```diff
@@ -0,0 +1,397 @@
+    if not IsExe(rpmbuild_path):
+      raise InvalidRpmbuildError('{} is not executable'.format(rpmbuild_path))
+    return rpmbuild_path
+  path = Which('rpmbuild')
+  if path:
+    return path
+  raise NoRpmbuildFoundError()
+
+
+class RpmBuilder(object):
+  """A helper class to manage building the RPM file."""
+
+  SOURCE_DIR = 'SOURCES'
+  BUILD_DIR = 'BUILD'
+  BUILDROOT_DIR = 'BUILDROOT'
+  TEMP_DIR = 'TMP'
+  RPMS_DIR = 'RPMS'
+  DIRS = [SOURCE_DIR, BUILD_DIR, BUILDROOT_DIR, RPMS_DIR, TEMP_DIR]
+
+  def __init__(self, name, version, release, arch, rpmbuild_path,
+               pre_scriptlet_path,
+               post_scriptlet_path,
+               preun_scriptlet_path,
+               postun_scriptlet_path,
+               debug=False):
+    self.name = name
+    self.version = GetFlagValue(version)
+    self.release = GetFlagValue(release)
+    self.pre_scriptlet = SlurpFile(pre_scriptlet_path) if pre_scriptlet_path is not None else ''
+    self.post_scriptlet = SlurpFile(post_scriptlet_path) if post_scriptlet_path is not None else ''
+    self.preun_scriptlet = SlurpFile(preun_scriptlet_path) if preun_scriptlet_path is not None else ''
+    self.postun_scriptlet = SlurpFile(postun_scriptlet_path) if postun_scriptlet_path is not None else ''
+    # TODO(nacl): delete this or properly integrate it into the call to
+    # rpmbuild(8).
+    self.arch = arch
+    self.files = []
+    self.rpmbuild_path = FindRpmbuild(rpmbuild_path)
+    self.rpm_path = None
+    self.debug = debug
+
+  def AddFiles(self, paths, root=''):
+    """Add a set of files to the current RPM.
+
+    If an item in paths is a directory, its files are recursively added.
+
+    Args:
+      paths: The files to add.
+      root: The root of the filesystem to search for files. Defaults to ''.
+    """
+    for path in paths:
+      full_path = os.path.join(root, path)
+      if os.path.isdir(full_path):
+        self.AddFiles(os.listdir(full_path), full_path)
+      else:
+        self.files.append(full_path)
+
+  def SetupWorkdir(self,
+                   spec_file,
+                   preamble_file,
+                   description_file,
+                   install_script_file,
+                   file_list_file,
+                   original_dir):
+    """Create the needed structure in the workdir."""
+
+    # Create directory structure.
+    for name in RpmBuilder.DIRS:
+      if not os.path.exists(name):
+        os.makedirs(name, 0o777)
+
+    # Copy the files.
+    for f in self.files:
+      dst_dir = os.path.join(RpmBuilder.BUILD_DIR, os.path.dirname(f))
+      if not os.path.exists(dst_dir):
+        os.makedirs(dst_dir, 0o777)
+      shutil.copy(os.path.join(original_dir, f), dst_dir)
+
+    # Copy the spec file, updating with the correct version.
+    spec_origin = os.path.join(original_dir, spec_file)
+    self.spec_file = os.path.basename(spec_file)
+    replacements = {}
+    replacements['BZL_PRE_SCRIPTLET'] = "%pre\n" + self.pre_scriptlet
```

Comment:
The BZL_ part seems unnecessary.

Reply:
Hm.  I put `BZL_` part in from to make it easier to distinguish between hand-added parts of the template and ones that will be filled out by rules_pkg itself.

I don't feel attached to it, though.  I'll look into this.

Reply:
Addressed in 229a244.